### ISSUE TITLE
Inconsistent behaviour of torch.fmod on CPU and GPU

### ISSUE THREAD

**Issue by nikitautiu** [pytorch/pytorch#15921]
## 🐛 Bug

`torch.fmod` yields different result between GPU and CPU for large numbers.

## To Reproduce
```python
import torch

print(torch.fmod(torch.tensor(421123111).cpu(), 2))  # 1
print(torch.fmod(torch.tensor(421123111).cuda(), 2))  # 0
```

## Expected behavior

Correct result for both runs or clear documentation about the inconsistency.

## Environment

```
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 410.79
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.3.1
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a

Versions of relevant libraries:
[pip3] numpy (1.15.4)
[pip3] torch (1.0.0)
[pip3] torchvision (0.2.1)
[conda] mkl                       2018.0.1             h19d6760_4  
[conda] mkl-service               1.1.2            py36h17a0993_4
```

**Comment by fmassa** [comment]
This is probably because in the GPU, we cast `fmod` to float and use `fmod` built-in when dealing with integer data
https://github.com/pytorch/pytorch/blob/517c7c98610402e2746586c78987c64c28e024aa/aten/src/THC/THCTensorMathPairwise.cu#L173-L183
while in the CPU we just use the `%` operator
https://github.com/pytorch/pytorch/blob/b28738ccb5b1cbf2c0ed3e6db5765aa4abc54ec3/aten/src/TH/generic/THTensorEvenMoreMath.cpp#L871

**Comment by igy254** [comment]
@fmassa I feel this would be a good start to contribute some code :) If this one is available I would love to solve it.

**Comment by fmassa** [comment]
Yes, go for it!

**Comment by chriscramer** [comment]
@igy254  I'm not sure how far you've gotten on this, but I'm 99% certain that the issue comes from the conversion to a single precision float when working on the GPU.  A single precision float only has 24 bits of mantissa, so it can only represent 24 bits of precision.  Any integer larger than that is going to be represented with less precision and a larger floating point exponent.  To test this theory, you can run:

    print(torch.fmod(torch.tensor(2**24 - 1).cuda(), 2))  # 1
    print(torch.fmod(torch.tensor(2**24 + 1).cuda(), 2))  # 0

You can also test with double precision and both values return a correct result (though of course would be wrong for values greater than 2^53)

It probably makes sense to have the CPU version also operate on floats so that they give the same values.  Failing that - I guess documentation

**Comment by fmassa** [comment]
Can't we have a templates function that uses `%` for integer types, as done in the GPU?

**Comment by chriscramer** [comment]
@fmassa I'm not very familiar with the codebase, but I think it boils down to a question of consistency.  How should fmod operate on integers?  

In C/C++, fmod casts an int/long to a double (or float for fmodf) which is also what seems to happen in pytorch on the GPU.  The inconsistency comes from pytorch using the integer modulo operator under the covers of the fmod function on ints.  It returns the mathematically correct answer, but not the answer that is functionally defined.

thoughts?

**Comment by fmassa** [comment]
My 2 cents in this case would be to use `%` for ints, but I'll also let the others chime in.
Cc @soumith

**Comment by igy254** [comment]
@chriscramer You are right about single precision float.

 If you take a look at IEEE 754 standard and use the number @nikitautiu gave as an example, you have the following:

` 0       10011011      10010001100111010100001
  +           (2**28)          1.5688058137893677 `

So we saved 421123104 and got the error of -7. 

I also agree that we should answer how `fmod()` should operate on integers. Conversion to float is not and option.

**Comment by igy254** [comment]
@fmassa  I guess the others had enough time to express them. Should I just submit my pull request as soon as I can?

**Comment by fmassa** [comment]
@igy254 I think using `%` for ints reasonable. I'm ccing @gchanan to see if he has a different view of it

**Comment by nikitautiu** [comment]
Seems sensible, although this being the issue which brought me here in the first place I guess `fmod`'s documentation should be updated to spare people future confusion.

**Comment by igor-susic** [comment]
@rohitrango Yes, well actually started again, as I did not have any GPU to test or develop with last few months, now brand new came from the USA and I can continue to work. Sorry everyone that I had not sent a notice to you all. By the way, the label `triaged` can be removed, sadly I can confirm that the issue is still there, tested it a few hours ago.

**Comment by rohitrango** [comment]
@igor-susic since you have been inactive for a while, I thought I'd give it a try.

**Comment by igor-susic** [comment]
@rohitrango No problem mate, better than waiting for me to find time for it :)

**Comment by fmassa** [comment]
@rohitrango note that this is going to be addressed in #24405 most probably

**Comment by sudo-panda** [comment]
Is this issue solved?

**Comment by mruberry** [comment]
This appears to be fixed on nightly, although it is not fixed in PyTorch 1.7. Maybe @ejguan fixed it?

**Comment by ejguan** [comment]
Yes, it's fixed. see PR #47323
